ircd/send.c

```diff
@@ -1503,3 +1503,139 @@ kill_client_serv_butone(struct Client *one, struct Client *target_p, const char
+		return false;
+	}
+
+	multiline_stashed_target_p = target_p;
+	multiline_cur_len = multiline_prefix_len;
+	multiline_cur = multiline_items;
+	return true;
+}
+
+enum multiline_item_result
+sendto_one_multiline_item(struct Client *target_p, const char *format, ...)
+{
+	va_list args;
+	char item[BUFSIZE];
+	int item_len, res;
+	enum multiline_item_result ret = MULTILINE_SUCCESS;
+
+	if (target_p != multiline_stashed_target_p)
+	{
+		multiline_stashed_target_p = NULL;
+		sendto_realops_snomask(SNO_GENERAL, L_ALL, "Multiline: missed init call!");
+		return MULTILINE_FAILURE;
+	}
+
+	va_start(args, format);
+	item_len = vsnprintf(item, sizeof item, format, args);
+	va_end(args);
+
+	if (item_len < 0 || (size_t)item_len >= sizeof item || (size_t)(multiline_prefix_len + item_len) > BUFSIZE - strlen("\r\n"))
+	{
+		multiline_stashed_target_p = NULL;
+		sendto_realops_snomask(SNO_GENERAL, L_ALL, "Multiline: failure preparing item! item_len: %d multiline_prefix_len: %d", item_len, multiline_prefix_len);
+		return MULTILINE_FAILURE;
+	}
+
+	if ((size_t)(multiline_cur_len + item_len) > BUFSIZE - strlen("\r\n"))
+	{
+		/* strip trailing space */
+		*(multiline_cur - 1) = '\0';
+		sendto_one(target_p, "%s%s", multiline_prefix, multiline_items);
+		multiline_cur_len = multiline_prefix_len;
+		multiline_cur = multiline_items;
+		ret = MULTILINE_WRAPPED;
+	}
+
+	res = snprintf(multiline_cur, sizeof multiline_items - (multiline_cur_len - multiline_prefix_len), "%s ", item);
+
+	if (res < 0)
+	{
+		multiline_stashed_target_p = NULL;
+		sendto_realops_snomask(SNO_GENERAL, L_ALL, "Multiline: failure appending item! res: %d", res);
+		return MULTILINE_FAILURE;
+	}
+
+	multiline_cur_len += res;
+	multiline_cur += res;
+	return ret;
+}
+
+bool
+sendto_one_multiline_fini(struct Client *target_p, const char *format, ...)
+{
+	va_list args;
+	char final[BUFSIZE];
+	int final_len;
+
+	if (target_p != multiline_stashed_target_p)
+	{
+		multiline_stashed_target_p = NULL;
+		sendto_realops_snomask(SNO_GENERAL, L_ALL, "Multiline: missed init call!");
+		return false;
+	}
+
+	if (multiline_cur_len == multiline_prefix_len)
+	{
+		multiline_stashed_target_p = NULL;
+		return true;
+	}
+
+	if (format)
+	{
+		va_start(args, format);
+		final_len = vsnprintf(final, sizeof final, format, args);
+		va_end(args);
+		if (final_len < 0 || (size_t)final_len >= sizeof final - strlen("\r\n") || final_len > multiline_prefix_len)
+		{
+			multiline_stashed_target_p = NULL;
+			sendto_realops_snomask(SNO_GENERAL, L_ALL, "Multiline: failure preparing final prefix! final_len: %d", final_len);
+			return false;
+		}
+	}
+	else
+	{
+		strncpy(final, multiline_prefix, sizeof final);
+		final_len = multiline_prefix_len;
+	}
+
+	/* strip trailing space */
+	*(multiline_cur - 1) = '\0';
+	sendto_one(target_p, "%s%s", final, multiline_items);
+
+	multiline_stashed_target_p = NULL;
+	return true;
+}
+
+void
+sendto_one_multiline_reset()
+{
+	multiline_stashed_target_p = NULL;
```

Comment:
Should this complain if a multiline was not in progress?

Reply:
I could see this go either way; so far the only usage is in CAP LS/LIST for < 3.2 clients, and not having this check for in-progress makes that implementation slightly simpler, because then it can just reset and return if result isn't success (no wrap), rather than reset only if wrapped, and then return if result isn't success.